### PR TITLE
Allow token replacement of .npmrc configuration with env vars

### LINKED ISSUE #1206: Environment variables are not replaced in .npmrc files
**Do you want to request a _feature_ or report a _bug_?**

Bug

**What is the current behavior?**

An .npmrc file like this:

//registry.npmjs.org/:_authToken=${NPM_TOKEN}

Yarn does not replace ${NPM_TOKEN} with the value of environment variable NPM_TOKEN.

**If the current behavior is a bug, please provide the steps to reproduce.**

See above.

**What is the expected behavior?**

Per https://docs.npmjs.com/files/npmrc, environment variable substitution is a feature expected to be used when parsing an .npmrc file by the client.

**Please mention your node.js, yarn and operating system version.**

yarn: 0.16.0
node: 6.9.0
os: Mac OS 10.12 (16A323)

### PR DESCRIPTION
**Pull request by darthtrevino**
**Summary**

NPM allows for clients to inject environment variables into .npmrc configuration files. e.g.

```
//registry.npmjs.org/:_authToken=${NPM_TOKEN}
registry=http://registry.npmjs.org
```

This commit allows yarn to understand .npmrc files using this feature.

**Test plan**
- [X] Build the yarn executable 
- [x] execute `yarn config list` to view injected variables

Fixes #1206

### CODE REVIEW

**Comment by darthtrevino** [comment]
dwapkyphuw

**Comment by holm** [comment]
We use this to inject our NPM token into the nvmrc file, so we can avoid checking in the token, and also avoid having to deal with global npm configuration on machines.

**Comment by samccone** [comment]
Thanks for the PR @darthtrevino I will review

**Comment by darthtrevino** [comment]
Sweet, thank you sir!

File: src/registries/npm-registry.js [hunk 1]
```diff
@@ -36,6 +36,24 @@ function getGlobalPrefix(): string {
   }
 }
 
+function envReplace(f: string): string {
```
Comment (wyze) [inline src/registries/npm-registry.js]: Should be more descriptive variable here than `f`.

**Comment by darthtrevino** [comment]
Word, thanks @wyze; in the last couple of commits I split out the envReplace function into a separate module in `utils/` and added a few basic tests

**Comment by maybeec** [comment]
Any chance to get this in the next release? Would love to see this as it is annoying me in my project :)

File: src/registries/npm-registry.js [hunk 2]
```diff
@@ -118,6 +119,9 @@ export default class NpmRegistry extends Registry {
 
     for (const [, loc, file] of await this.getPossibleConfigLocations('.npmrc')) {
       const config = ini.parse(file);
+      Object.keys(config).forEach((key: string) => {
```
Comment (kittens) [inline src/registries/npm-registry.js]: Use `for...in` here please.

File: __tests__/util/env-replace.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/* @flow */
+import envReplace from '../../src/util/env-replace';
+import assert from 'assert';
+
+describe('environment variable replacement', () => {
+
```
Comment (kittens) [inline __tests__/util/env-replace.js]: Remove empty newline.

File: src/util/env-replace.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+/* @flow */
+export default function envReplace(value: string, env: {[key: string]: ?string} = process.env): string {
+  if (typeof value !== 'string' || !value) {
+    return value;
+  }
+
+  const envExpr = /(\\*)\$\{([^}]+)\}/g;
```
Comment (kittens) [inline src/util/env-replace.js]: Put this at the top level and capitalise it as `ENV_EXPR`.

**Comment by darthtrevino** [comment]
The CircleCI issue looks like it's affecting newer PRs too. Let me know when I can pull a fix for it.

**Comment by maybeec** [comment]
Does this PR also include the ability to replace any env variable? NPM, e.g., does not replace env vars in keys like this:

```
@scope:registry=${ARTIFACTORY_REGISTRY}
//${ARTIFACTORY_REGISTRY}:_password=${PASSWORD}
//${ARTIFACTORY_REGISTRY}:username=${USERNAME}
```

So ${ARTIFACTORY_REGISTRY} is replaced in a value expression but not in a key expression, which is quite odd.

**Comment by darthtrevino** [comment]
@maybeec Nope, this change acts the same way that NPM does. It performs token replacement in value expressions

**Comment by maybeec** [comment]
@darthtrevino maybe not in near future: npm/npm#14513 
So you might want to think of supporting this directly. :+1:

**Comment by darthtrevino** [comment]
@maybeec When NPM changes, I'll gladly implement it here (provided this PR actually makes it in)

**Comment by darthtrevino** [comment]
@kittens The changes you requested have been implemented and the CI looks good now.

**Comment by PAkerstrand** [comment]
Does this PR require further work/changes? Would love to see it merged...

**Comment by darthtrevino** [comment]
¯\_(ツ)_/¯  - I've addressed the review issues; it's up to the yarn team now

**Comment by darthtrevino** [comment]
💃

**Pull request merged**